Upgrade webpack

Timer opened the pull request:
webpack has had its first v2 release! 🎉

There are a few changes we're waiting on:
- [x] [babel-loader@6.2.10](https://github.com/babel/babel-loader/releases/tag/v6.2.10)
- [x] [webpack-dev-server@2.2.0-rc.0](https://github.com/webpack/webpack-dev-server/releases/tag/v2.2.0-rc.0)
- [x] https://github.com/ampedandwired/html-webpack-plugin/pull/526
- [x] [extract-text-webpack-plugin#309](https://github.com/webpack/extract-text-webpack-plugin/issues/309)
- [x] remove LoaderOptionsPlugin for prod postcss
  - [x] https://github.com/webpack/extract-text-webpack-plugin/issues/265
  - [x] https://github.com/webpack/extract-text-webpack-plugin/issues/281
- [ ] Check up on https://twitter.com/simplesthing/status/823674445843734528 – is this affecting us?
- [ ] https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/294
- [ ] https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/366

/cc #183

bondz commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -163,44 +163,24 @@ module.exports = {
       {
         test: /\.json$/,
-        loader: 'json'
+        loader: 'json-loader'
```
json-loader is no longer required in Webpack 2

Timer replied:
Thanks!

bondz replied:
You should take it out of the package.json file as well.

bondz commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -81,12 +81,12 @@ module.exports = {
     // We use `fallback` instead of `root` because we want `node_modules` to "win"
```
Should the comment change?

Timer replied:
Probably -- we don't need people to be confused.

bondz commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -134,7 +134,7 @@ module.exports = {
         ],
-        loader: 'url',
+        loader: 'url-loader',
         query: {
```
Although webpack 2 allows `query`, it has chosen to use `options` in its docs which IMO is better, would you consider changing?

bondz commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -232,5 +232,11 @@ module.exports = {
+  // splitting or minification in interest of speed. These warnings become
+  // cumbersome.
+  performance: {
+    hints: false
```
Since this is turned off completely during development and by default set to `warning` in the production config, it might come as a surprise to users to then have to tweak their app for code-splitting, although extracting css should bring the file size down. @TheLarkInn gives a reason to [leave it enabled](/webpack/webpack.js.org/pull/500/files#r92665821) in dev builds. We could tweak the `maxAssetSize` and `maxEntrypointSize` options to reasonable file sizes instead. What do you think?

Timer replied:
I don't think it's very valuable in its certain form -- I think we need to go a step further and show our own warnings more in line with what @TheLarkInn was saying.

Currently, we don't display build size warnings -- and bundle size is going to vary wildly between end-users. I don't think there is a good "default" we can set which will cover all cases.
Knowing this, I think that monitoring build size between rebuilds would be beneficial to let the user know their bundle has grown -- maybe doing some caching so we can display warnings between restarts/sessions.

Since the feature isn't there now, I think we can safely disable it completely until we implement said feature or wait for more detailed [json] performance hints from webpack.

bondz replied:
I agree 👍

existentialism replied:
Just a note, performance hints are now off by default in rc4:
https://github.com/webpack/webpack/releases/tag/v2.2.0-rc.4

bondz replied:
Should this be set to `warning` in production now?

TheLarkInn replied:
What kind of additional details are you looking for? At the least I'd recommend having it as "warning" in Production as well.

bondz commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -163,44 +163,18 @@ module.exports = {
-      {
-        test: /\.json$/,
-        loader: 'json'
+        loader: 'style-loader!css-loader?importLoaders=1!postcss-loader'
```
To apply multiple loaders, webpack 2 uses the `use` option.

Timer replied:
Out of curiosity, is this old syntax still supported?

gaearon replied:
Nice! (That syntax is so confusing.)

bondz replied:
use 
```js
use: [
  'style-loader', 
  {
    loader: 'css-loader',
    options: {
      importLoaders: 1
    }
  },
  'postcss-loader'
]
```

existentialism commented:
Probably need to bump a few deps for peerDep warnings:

[babel-loader@6.2.10](https://github.com/babel/babel-loader/releases/tag/v6.2.10)
[webpack-dev-server@2.2.0-rc.0](https://github.com/webpack/webpack-dev-server/releases/tag/v2.2.0-rc.0)

And wait for some pending version bumps:

[html-webpack-plugin#525](https://github.com/ampedandwired/html-webpack-plugin/issues/525)
[extract-text-webpack-plugin#309](https://github.com/webpack/extract-text-webpack-plugin/issues/309)

bondz commented on packages/react-scripts/config/webpack.config.prod.js:
```diff
@@ -172,47 +172,22 @@ module.exports = {
-        loader: ExtractTextPlugin.extract('style', 'css?importLoaders=1!postcss')
+        loader: ExtractTextPlugin.extract({
+          fallbackLoader: 'style-loader',
+          loader: 'css-loader?importLoaders=1!postcss-loader'
```
This could also be written as

```js
loader: [
            {
              loader: 'css-loader',
              options: {
                importLoaders: 1
              }
            },
            'postcss-loader'
          ]
```

bondz commented:
Some other dependencies are outdated, should those deps be updated as well? In a separate PR, I could send one.

Timer commented:
> Some other dependencies are outdated, should those deps be updated as well? In a separate PR, I could send one.

I've always wondered the same and if we should track these dependencies through their new versions.
I feel like the general consensus is if it works, leave it as it is. If we update dependencies we might introduce breakages.
To update dependencies safely, we need extensive unit testing beyond the current smoke test -- this is on the horizon.

wtgtybhertgeghgtwtg commented:
Would that mean enabling Greenkeeper after #1187 lands?

Timer commented:
We'll have to open a discussion issue about it -- I can think of plenty of reasons to argue for and against it.
However, with an appropriate test suite, I don't believe it could hurt.

---

Anyway, this PR is currently blocked on waiting for a few dependencies to update. Hopefully webpack 2.x is fully released by this time. 😄

kpuputti commented:
The linked [html-webpack-plugin PR](https://github.com/ampedandwired/html-webpack-plugin/pull/526) is now merged and released, so that dependency can be checked. \o/

Timer commented:
Neat, thanks for the update @kpuputti!

existentialism commented:
Can bump to [extract-text-plugin@v2.0.0-beta.5](https://github.com/webpack/extract-text-webpack-plugin/releases/tag/v2.0.0-beta.5) now! 🙌

kirkaustin commented:
Webpack 2.2.0 has been released.

Timer commented:
Thanks @existentialism -- I've added the biggest two blockers for us to the OP.

---

Thanks @kirkaustin, I've updated this PR accordingly. 👍

gaearon commented:
Tagging this for 0.10 together with other major features.

SpaceK33z commented on packages/react-scripts/package.json:
```diff
@@ -43,25 +43,24 @@
-    "webpack": "1.14.0",
-    "webpack-dev-server": "1.16.2",
+    "webpack": "2.2.0",
+    "webpack-dev-server": "2.2.0-rc.0",
```
webpack-dev-server `2.2.0` was released[ a few days ago](https://github.com/webpack/webpack-dev-server/releases/tag/v2.2.0) :).

sashakru commented:
This two issues are closed:

webpack-contrib/extract-text-webpack-plugin#265
webpack-contrib/extract-text-webpack-plugin#281

bachstatter commented on packages/react-scripts/config/webpack.config.prod.js:
```diff
@@ -172,47 +183,30 @@ module.exports = {
+          loader: [
+            {
+              loader: 'css-loader',
+              query: {
```
query -> options?

Timer replied:
That change is pending on https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/281, which I believe has gone through.

gaearon commented:
Is anything still blocking this?

bondz commented:
Looks like all known issues have been resolved. Just a few more suggested changed by reviewers.

gaearon commented:
>But extract-text-webpack-plugin seems to have some bugs with webpack 2.2, we should probably wait for it's stable release.

👍 

Can you compile a list of known issues for 2.x?

bondz commented:
It looks like the api would change in the stable release - webpack-contrib/extract-text-webpack-plugin#382

Also, these could affect some users. 
- webpack-contrib/extract-text-webpack-plugin#307
- webpack-contrib/extract-text-webpack-plugin#294

Timer commented:
I'll sit down and bring this PR up to date tonight.

Timer commented:
Should be updated! Let me know if anyone sees anything abnormal.

----

> Check up on https://twitter.com/simplesthing/status/823674445843734528 – is this affecting us?

There has been quite a few releases since that was posted which addressed issues we were having (with configuration), but I'm not sure if any were pertaining to private repos. She continues to say she identified the issue, so it might have been a configuration one.

gaearon commented:
Hmm, CI is still failing :-(

Timer commented:
@gaearon I believe it has to do with symlinking ... The root cause is so puzzling (things are only blowing up due to unfound deps).

Perhaps due to this (if you noticed, installs work fine): https://github.com/lerna/lerna/commit/143840eac0117e15c3a9180d8cc48ad830d34f47
https://github.com/lerna/lerna/issues/11
?

Rouby commented:
Will Tree-Shaking work with this configuration?

gaearon commented:
We really need someone to look into the CI failure (perhaps, separately from this PR). It’s annoying we can’t see if changes work when they add dependencies. There’s got to be some hack we can add, even if it’s a Lerna bug.

Timer commented:
webpack-dev-server@2.3.0 was released and added our error overlay!
https://github.com/webpack/webpack-dev-server/commit/d725fd9f3b0c2ab92e6489c993f80602c0e47848
😮

Should we disable it in the options or remove ours, @gaearon?

Timer commented:
@EnoahNetzach any idea about this failure?
https://travis-ci.org/facebookincubator/create-react-app/jobs/198416336#L1594

SpaceK33z commented:
@Timer, haha yes I copied the error overlay :). OSS <3. It's probably too early for CRA to switch to it. It doesn't have support yet for warnings and doesn't use a formatter [like you built](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-dev-utils/formatWebpackMessages.js). Perhaps on long term we could built the right hooks into dev-server?

EnoahNetzach commented:
@Timer I used to have it only ofter an unsuccesful/inconsistent eject.
It could have something to do with devdep not getting installed (another recurring problem I have somewhere else).

I'll try to get some time to try it locally later this evening.

Timer commented:
@EnoahNetzach does it have to do with the fact you're removing the `.babelrc` for the ejected test? I thought we disregarded `.babelrc` anyway, how did you get it to work?

Timer commented:
@bondz 
I don't think https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/307 affects us because we only use `extract-text-webpack-plugin` in production, not development, so watch mode isn't utilized in conjunction with the plugin.

There has been no movement on https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/294 so I think we're going to need to figure out a reproducing case and fix it ourselves / pass off to @bebraw.

bondz commented:
@Timer true,  webpack-contrib/extract-text-webpack-plugin#307 doesn't affect CRA for stated reason, glad. 

Think  webpack-contrib/extract-text-webpack-plugin#366 might affect us though but the issues don't seem to have test cases the maintainers need for a fix and could take a while for them to be provided. 

Guess we can ask @bebraw for a recommendation.

bebraw commented:
@bondz Yeah, we are a little stuck with a couple of issues. Those prevent from going "final" with the plugin. If the situation keeps like this, I just might have to call it stable even without those two fixes.

Timer commented:
CI passing 🎉

Timer commented:
🚀 let's do it!

gaearon commented:
![](https://media.giphy.com/media/3osxYdwmc9Oy8H7h7i/giphy.gif)

bryant988 commented on packages/react-scripts/package.json:
```diff
@@ -61,8 +60,8 @@
+    "webpack-dev-server": "2.3.0",
     "webpack-manifest-plugin": "1.1.0",
     "whatwg-fetch": "2.0.2"
   },
```
Should we include the exclusion of "/scripts" folder in jest ignore patterns here since it tries to run `test.js`?

```"testPathIgnorePatterns": [
      "<rootDir>[/\\\\](build|docs|node_modules|scripts)[/\\\\]"
],```

sublimeye commented:
@Timer Is this change should be already available when installing a new app? I'm still getting webpack@1.14 with a new app.

Created a bug just in case #1599

gaearon commented:
Just because something was merged, doesn't mean it was released 😉 .

Timer commented:
No @sublimeye, it is only available in master and must be installed manually.

gaearon commented:
Just to be clear, I wouldn't recommend installing it manually to anyone. We're still working out the issues, and master is never considered fully stable.

You can see the milestone for each issue and PR:

<img width="284" alt="screen shot 2017-02-20 at 18 03 38" src="https://cloud.githubusercontent.com/assets/810438/23136911/0424f81c-f797-11e6-9078-cd5fea8750a8.png">

Compare it to the [latest release](https://github.com/facebookincubator/create-react-app/releases) to know if something is released or not.

shai32 commented:
When will 0.10 will be released?

SimenB commented on packages/react-scripts/package.json:
```diff
@@ -39,20 +39,19 @@
     "eslint-plugin-jsx-a11y": "2.2.3",
     "eslint-plugin-react": "6.4.1",
-    "extract-text-webpack-plugin": "1.0.1",
+    "extract-text-webpack-plugin": "2.0.0-rc.3",
```
2.0.0 has been released

https://github.com/webpack-contrib/extract-text-webpack-plugin/releases/tag/v2.0.0

gaearon commented:
@shai32 

When everything [tagged for 0.10](https://github.com/facebookincubator/create-react-app/milestone/23) is ready. We don’t delay releases because we want to—we delay them because there are outstanding bugs (in CRA or underlying tools), or because there are features we still want to get in. Maybe you want to help?

shai32 commented:
Sorry for asking for release date, will not do it again.
Thanks for anyone who contribute to this project.

gaearon commented:
No worries, I’m sorry if my message sounded combative 😞 .

I just meant that we don’t really have any specific day in mind—it’ll take some time to fix all the issues, and help is welcome.

iddan commented:
For some reason even though I upgraded the package it still installs me webpack 1

gaearon commented:
Have you had a chance to read the comments above asking about the very same question?
😉

viksit commented:
@gaearon how do you recommend installing CRA from github? I've tried `npm install -g git+https://github.com/facebookincubator/create-react-app.git` but that results in a 

```
npm ERR! No name provided in package.json

npm ERR! A complete log of this run can be found in: ...
```

What am I doing wrong?

gaearon commented:
We don't support installing from GitHub (and that won't help you anyway because the CLI is not where the code lives).

Please don't rush to migrate to Webpack 2 because it's cool. We'll release support for it when it's fully ready. Give it some time please.

If you absolutely insist on trying it, there's an **unstable, possibly buggy, nightly canary build** of `react-scripts` that includes Webpack 2. I will lock this thread and won't tell you how to find it because if it's posted here, then every beginner will attempt to use it, and their projects will break in unexpected ways. But if you know how to find unstable npm package versions, then feel free to give it a try and report any issues 😉 .

gaearon commented:
Please help beta test the new version that includes this change!
https://github.com/facebookincubator/create-react-app/issues/2172
